### ISSUE TITLE
5$: Span for E064 is too much

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#280]
In the following code, quick-lint-js has a very big span for the E064 error:
```js
{
    if (isNaN(port))
//  ^^^^^^^^^^^^^^^^ E064 missing body for 'if' statement
}
```
Make the error point to after the ')' instead (i.e. where the body would be expected).

Perhaps include a note of where the `if` statement begins.

**Comment by AidenThing** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.
Hey I would like to do this already started a little bit so a question:
do you want squiggles after the 'if(a)' statement if there is no '\n' after?
wouldn't it imply that the user is still typing. also, it still throws error if you call quick-lint-js exampltxt.mjs. just no squiggles

**Comment by strager** [comment]
>  do you want squiggles after the 'if(a)' statement if there is no '\n' after?

If there's no newline in `{if(a)}`, I want quick-lint-js to report an empty span between the `)` and the `{`.

> wouldn't it imply that the user is still typing.

An error is an error.

I do want to tailor the error messaging to be as helpful as possible, which is the point of this task. I think the current behavior (highlighting the whole `if(condition)`) is noisy and unhelpful.

> also, it still throws error if you call quick-lint-js exampltxt.mjs. just no squiggles

The command line interface doesn't show squigglies right now. You'd need to use the Vim plugin or VS Code extension (or the [web demo](https://quick-lint-js.com/demo/), which is kind of jank right now) to see any squigglies.

**Comment by AidenThing** [comment]
Hey, is the the kind of formatting you are looking for?
Would you want, in case d for it to not continue to the next line between the if and the else? just want it to stop on the line of the if?

Also, I do not know what to do about quick-lint calling error 150 instead on line 6. I think that that is a separate problem.
 
![quick-lint](https://user-images.githubusercontent.com/60998375/118177964-01998900-b3f9-11eb-9d1b-c68a158b09a3.PNG)

**Comment by strager** [comment]
> Hey, is the the kind of formatting you are looking for?
>
> ![quick-lint](https://user-images.githubusercontent.com/60998375/118177964-01998900-b3f9-11eb-9d1b-c68a158b09a3.PNG)

* First **a** looks good assuming only line 2 has a squiggly (not like 3).
* Second **a** looks good.
* **b** looks good, but it'd be better for the squiggly to not cover the `}`.
* **c** looks good.
* **d** looks good assuming only line 17 has a squiggly. If line 18-19 are because of quick-lint-js, that's a problem.

> Would you want, in case d for it to not continue to the next line between the if and the else? just want it to stop on the line of the if?

Correct.

> Also, I do not know what to do about quick-lint calling error 150 instead on line 6. I think that that is a separate problem.

E150 is similar to E064. I'd say this task covers both.

E150 is like E064 but includes a 'note'. Notes are currently not visible in VS Code (#200). It is visible in the CLI though:
```js
if (true)
let after;
```
```
$./build/quick-lint-js hello.js
hello.js:1:10: error: missing body for 'if' statement [E150]
hello.js:2:1: note: a lexical declaration is not allowed as the body of an 'if' statement [E150]
```

**Comment by AidenThing** [comment]
Hey, if an {if(a)} is followed by another {if(a)} only the second {if(a)} will get the squiggles [img {1}]. The same thing happens on the web view [img {2}].
I don't know where this error is coming from, but I haven't seen any other posts about and just thought it should be on your radar.

img 1
![qlthing](https://user-images.githubusercontent.com/60998375/118218123-6b875200-b43c-11eb-943e-94ff5381972f.PNG)

img 2
![qlqeb](https://user-images.githubusercontent.com/60998375/118218459-10a22a80-b43d-11eb-9939-77054201f27b.PNG)

However, aside from that here is what I have so far,
(Sorry that I have to label which squiggles are from quick-lint and not from quick-lint. Something in vscode is doing it automatically and I can't turn it off )

![quick-lint-sec](https://user-images.githubusercontent.com/60998375/118219064-57dceb00-b43e-11eb-93b3-8d1879900d69.PNG)

I think I have it working well, the only thing would be where the squiggle lies after the {if(a)} if there is a space or not.
If this is up to snuff I would be ready to send it.

However, if it is not, I honestly would not know how to make the squiggle extend after the{if(a)} if there is not space as whitespace is mostly ignored.

**Comment by strager** [comment]
> Hey, if an {if(a)} is followed by another {if(a)} only the second {if(a)} will get the squiggles [img {1}]. The same thing happens on the web view [img {2}].
> I don't know where this error is coming from, but I haven't seen any other posts about and just thought it should be on your radar.
> 
> img 1
> ![qlthing](https://user-images.githubusercontent.com/60998375/118218123-6b875200-b43c-11eb-943e-94ff5381972f.PNG)
> 
> img 2
> ![qlqeb](https://user-images.githubusercontent.com/60998375/118218459-10a22a80-b43d-11eb-9939-77054201f27b.PNG)

This doesn't look like a problem to me at all. The following code is 100% legal:
```js
if (a)
  if (b)
    c();
```

If you remove `c();`, a naive parser (like quick-lint-js') will think that `if (b)` is missing its body. `if (a)` has a body, though, which is `if (b)`, so there's no error for `if (a)`.

> However, aside from that here is what I have so far,
> (Sorry that I have to label which squiggles are from quick-lint and not from quick-lint. Something in vscode is doing it automatically and I can't turn it off )
> 
> ![quick-lint-sec](https://user-images.githubusercontent.com/60998375/118219064-57dceb00-b43e-11eb-93b3-8d1879900d69.PNG)
> 
> I think I have it working well, the only thing would be where the squiggle lies after the {if(a)} if there is a space or not.
> If this is up to snuff I would be ready to send it.

Looks great!

> However, if it is not, I honestly would not know how to make the squiggle extend after the{if(a)} if there is not space as whitespace is mostly ignored.

It might be an issue with our VS Code extension, not with the core of quick-lint-js.
